x/subscription/spec/03_messages.md

```diff
@@ -0,0 +1,94 @@
+<!--
+order: 3
+-->
+
+# Messages
+
+## MsgCreatePlan
+
+Merchant or payment provider create plans.
+
+```protobuf
+message MsgCreatePlan {
+  string title;
+  string description;
+  Decimal price,  // price in fiat currency show to user
+  string fiat_currency,  // fiat currency name
+  string token_denom,  // denomination of token
+  Decimal conversion_rate,  // conversion rate from fiat currency to token, maintained by owner
+  int32 interval
+}
+```
+
+The message signer is the owner.
+
+## MsgStopPlan
+
+```protobuf
+message MsgStopPlan {
+  int32 planId = 1,
+}
+```
+
+Can only stopped by the owner of the plan, also removes all subscriptions.
+
+## MsgChangeConversionRate
+
+```protobuf
+message MsgChangeConversionRate {
+  int32 planId = 1,
+  Decimal rate = 2,
+}
+```
+
+Can only be changed by plan owner.
+
+## MsgCreateSubscription
+
+```protobuf
+message MsgCreateSubscription {
+  int32 planId = 1,
+}
+```
+
+The message signer subscribe the plan.
+
+## MsgStopSubscription
+
+```protobuf
+message MsgStopSubscription {
+  int32 planId = 1,
+  Address subscriber = 2,
+}
+```
+
+The subscriber unsubscribe the plan, or plan owner cancel a subscription.
+
+## MsgCollectPayments
+
+```protobuf
+message MsgCollectPayments {
+  int32 planId = 1,
+}
+```
+
+Can only issued by plan owner.
+
+`x/subscription` module automatically collect payments from subscribers at begin block, and write collection results into events. Merchants or payment providers can watch the events, if there's failed payment, they can choose to stop the off-chain service or cancel the subscription.
```

Comment:
for the automatic collection (without a manual tx by the plan owner), one thing to consider is how this could be misused -- as e.g. one could create plan-subscription cycles every second.
so one would either need to deduct fee for each operation or rate-limit the operations -- e.g. force that the plan has expiration time + there's a network param for "max collections"... so when creating the plan, one validation will be "(expiration time - time now) / period <= max collections".

one other random thought I had is that this could potentially be generalised not only for "msgsend", but for other tx that one may want to do periodically (withdraw reward, delegate...) -- but that's perhaps more advanced than the original goal / purpose of this module

Reply:
sounds like a standalone `cron` module ;D

Reply:
Weave (alternative to Cosmos SDK) has this module: https://github.com/iov-one/weave/tree/master/x/cron :D

Reply:
but that was just a random thought -- solving the original problem of payment subscriptions is useful on its own and generalising it may make it more complex than it needs to be

Reply:
Regarding the rate-limiting, I think we can consume some gas(https://docs.cosmos.network/master/basics/gas-fees.html) for CreatePlan message to cover the potential calculations later on.

Reply:
the cost of the plan will depend on how many subscriptions there are -- so gas for creating subscription can take into account how many executions it'll lead to depending on the plan expiration and period?
given the expected initial gas price levels + exchange rate, not returning the "unused" gas from cancelled subscriptions shouldn't be a big issue for now -- similarly to what @christopher-crypto mentioned regarding the manual "retry" gas costs

Reply:
Added this:
```
period0 = (block_time + interval_offset) / interval
periodn = (expiration_time + interval_offset) / interval
ConsumeGas( (periodn - period0 + 1) * GasPerSubscription )
```

`GasPerSubscription` is a network parameter.